spec.bs

```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
 <div algorithm>
   To <dfn>canonicalize protocol</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. Let |strippedValue| be the given |value| with a single trailing U+003A (:) removed, if any.
+  1. If |type| is "pattern" then return |strippedValue|.
+  1. Let |dummyURL| be a new [=url=] record.
+  1. Let |parseResult| be the result of [=Basic URL parser|URL parsing=] |strippedValue|, followed by U+003A (:), with |dummyURL| as [=basic URL parser/url=].
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/scheme=].
 </div algorithm>
 
 <div algorithm>
   To <dfn>canonicalize username</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. If |type| is "pattern" then return |value|.
+  1. Let |dummyURL| be a new [=url=] record.
+  1. [=Set the username=] given |dummyURL| and |value|.
+  1. Return |dummyURL|'s [=url/username=].
 </div algorithm>
 
 <div algorithm>
   To <dfn>canonicalize password</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. If |type| is "pattern" then return |value|.
+  1. Let |dummyURL| be a new [=url=] record.
+  1. [=Set the password=] given |dummyURL| and |value|.
+  1. Return |dummyURL|'s [=url/password=].
 </div algorithm>
 
 <div algorithm>
   To <dfn>canonicalize hostname</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. If |type| is "pattern" then return |value|.
+  1. Let |dummyURL| be a new [=url=] record.
+  1. Let |parseResult| be the result of [=Basic URL parser|URL parsing=] |value| with |dummyURL| as [=basic URL parser/url=] and [=hostname state=] as [=basic URL parser/state override=].
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/host=].
 </div algorithm>
 
 <div algorithm>
-  To <dfn>canonicalize port</dfn> given a string |value| and a string |type|:
-
-  1. TODO |value| |type|
+  To <dfn>canonicalize port</dfn> given a string |portValue|, a string |protocolValue|, and a string |type|:
```

Comment:
```suggestion
  To <dfn>canonicalize port</dfn> given a string |portValue|, a string-or-null |protocolValue|, and a string |type|:
```

Reply:
This algorithm's current call site only gives one argument. I think it should explicitly give null for the second.

Reply:
This was a mistake.  It should be passing a protocol.